### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by sergiy17**
gem 'devise'
gem 'omniauth-facebook'
Facebook APP - done
page creation - done

- [ ] @vovka 
- [x] @brouhaha88 
- [x] @superNazo

### CODE REVIEW

**Comment by anatoliiibrahimov** [comment]
- [ ] @vovka 
- [ ] @brouhaha88 
- [ ] @superNazo

File: config/initializers/devise.rb [hunk 1]
```diff
@@ -0,0 +1,265 @@
+# Use this hook to configure devise mailer, warden hooks and so forth.
+# Many of these configuration options can be set straight in your model.
+Devise.setup do |config|
```
Comment (brouhaha88) [inline config/initializers/devise.rb]: Please remove unnecessary coments.

File: spec/models/user_spec.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
```
Comment (brouhaha88) [inline spec/models/user_spec.rb]: Reason Coveralls "complains" is here..

File: config/initializers/devise.rb [hunk 3]
```diff
@@ -7,13 +7,13 @@
...
   # by default. You can change it below and use your own secret key.
   # config.secret_key = '5f97720cc4a4e125b8d67d26698be271e33ef72065ce92dd218645d98dfb821915566e7de346be933cc2635a3fb1dbb1cddd51d4d79b631c5f15509922a187bd'
-  config.omniauth :facebook, 'APP_ID', 'APP_SECRET', scope: 'user,public_repo'
   # ==> Mailer Configuration
   # Configure the e-mail address which will be shown in Devise::Mailer,
   # note that it will be overwritten if you use your own mailer class
   # with default "from" parameter.
-  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
-
+  config.omniauth :facebook, "1706833466223396", "fe92b6e4f3dd0ef2fefe15d63f92d395",
```
Comment (brouhaha88) [inline config/initializers/devise.rb]: Is it ok for this key to be here ?

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -1,8 +1,11 @@
 class ApplicationController < ActionController::Base
-  # Prevent CSRF attacks by raising an exception.
-  # For APIs, you may want to use :null_session instead.
-  protect_from_forgery with: :exception
```
Comment (brouhaha88) [inline app/controllers/application_controller.rb]: We don't need CSRF protection?

File: config/initializers/devise.rb (hunk 1 above)
Comment (vovka) [inline config/initializers/devise.rb]: @brouhaha88 you're right, we shouldn't leave comments. but these ones are generated by devise and give us some instructions and act as documentation. So i believe it's not critical to leave exactly these comments.

File: config/initializers/devise.rb (hunk 3 above)
Comment (vovka) [inline config/initializers/devise.rb]: I agree, we need to move the keys to move either to yaml file or to system's env vars.

File: app/controllers/application_controller.rb (hunk 4 above)
Comment (vovka) [inline app/controllers/application_controller.rb]: We definitely require it.

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -1,8 +1,11 @@
...
-  def hello
-  	render text: "hi from Mars!"
+  before_action :configure_permitted_parameters, if: :devise_controller?
+
+  protected
+
+  def configure_permitted_parameters
+  devise_parameter_sanitizer.for(:sign_up) { |u| u.permit(:username, :email, :password, :password_confirmation, :remember_me) }
```
Comment (vovka) [inline app/controllers/application_controller.rb]: What kinds of authentication should we have in our app? 
Do we really need username, login, email, password etc fields?

File: app/controllers/home_controller.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class HomeController < ApplicationController
```
Comment (vovka) [inline app/controllers/home_controller.rb]: Do we really need this controller?

File: app/controllers/users/home_controlle.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+class HomeController < ApplicationController
```
Comment (vovka) [inline app/controllers/users/home_controlle.rb]: Do we really need this controller?

File: app/controllers/home_controller.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
...
+  def index
+
+  end
+end
```
Comment (vovka) [inline app/controllers/home_controller.rb]: Please, double check an empty line in the end of file is present.

File: app/controllers/users/home_controlle.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
...
+  def index
+
+  end
+end
```
Comment (vovka) [inline app/controllers/users/home_controlle.rb]: Please, double check an empty line in the end of file is present.

File: app/controllers/users/home_controlle.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+class HomeController < ApplicationController
+  def index
+
```
Comment (vovka) [inline app/controllers/users/home_controlle.rb]: Redundant empty line

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+    else
+      session["devise.facebook_data"] = request.env["omniauth.auth"]
+      redirect_to new_user_registration_url
+    end
+  end
+
+  def failure
+    redirect_to root_path
+  end
+end
```
Comment (vovka) [inline app/controllers/users/omniauth_callbacks_controller.rb]: Please, double check an empty line in the end of file is present.

File: app/models/user.rb [hunk 12]
```diff
@@ -0,0 +1,16 @@
+class User < ActiveRecord::Base
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable
+  
+  devise :omniauthable, :omniauth_providers => [:facebook]
+  def self.from_omniauth(auth)
+  where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
```
Comment (vovka) [inline app/models/user.rb]: Wrong indentation, missed 1 level indent.

File: app/models/user.rb [hunk 13]
```diff
@@ -0,0 +1,16 @@
+class User < ActiveRecord::Base
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable
+  
+  devise :omniauthable, :omniauth_providers => [:facebook]
```
Comment (vovka) [inline app/models/user.rb]: `devise` appears twice, please make it only once.

File: app/views/devise/confirmations/new.html.erb [hunk 14]
```diff
@@ -0,0 +1,16 @@
+<h2>Resend confirmation instructions</h2>
```
Comment (vovka) [inline app/views/devise/confirmations/new.html.erb]: We don't need confirmation.

File: app/views/devise/mailer/confirmation_instructions.html.erb [hunk 15]
```diff
@@ -0,0 +1,5 @@
+<p>Welcome <%= @email %>!</p>
```
Comment (vovka) [inline app/views/devise/mailer/confirmation_instructions.html.erb]: We don't need confirmation instructions

File: app/views/devise/mailer/password_change.html.erb [hunk 16]
```diff
@@ -0,0 +1,3 @@
+<p>Hello <%= @resource.email %>!</p>
```
Comment (vovka) [inline app/views/devise/mailer/password_change.html.erb]: Since we are going to use soc networks auth only, we don't need passwords

File: app/views/devise/mailer/reset_password_instructions.html.erb [hunk 17]
```diff
@@ -0,0 +1,8 @@
+<p>Hello <%= @resource.email %>!</p>
```
Comment (vovka) [inline app/views/devise/mailer/reset_password_instructions.html.erb]: Since we are going to use soc networks auth only, we don't need passwords
 Add a line note

File: app/views/devise/mailer/unlock_instructions.html.erb [hunk 18]
```diff
@@ -0,0 +1,7 @@
+<p>Hello <%= @resource.email %>!</p>
```
Comment (vovka) [inline app/views/devise/mailer/unlock_instructions.html.erb]: We don't need locking/unlocking

File: app/views/devise/passwords/edit.html.erb [hunk 19]
```diff
@@ -0,0 +1,25 @@
+<h2>Change your password</h2>
```
Comment (vovka) [inline app/views/devise/passwords/edit.html.erb]: We don't need passwords at all

File: app/views/devise/passwords/new.html.erb [hunk 20]
```diff
@@ -0,0 +1,16 @@
+<h2>Forgot your password?</h2>
```
Comment (vovka) [inline app/views/devise/passwords/new.html.erb]: We don't need passwords at all

File: app/views/devise/registrations/edit.html.erb [hunk 21]
```diff
@@ -0,0 +1,39 @@
+<h2>Edit <%= resource_name.to_s.humanize %></h2>
```
Comment (vovka) [inline app/views/devise/registrations/edit.html.erb]: We don't need to edit account

File: app/views/devise/registrations/new.html.erb [hunk 22]
```diff
@@ -0,0 +1,29 @@
+<h2>Sign up</h2>
```
Comment (vovka) [inline app/views/devise/registrations/new.html.erb]: We don't need email, password etc. We only need authentication via FB.

File: app/views/devise/sessions/new.html.erb [hunk 23]
```diff
@@ -0,0 +1,26 @@
+<h2>Log in</h2>
```
Comment (vovka) [inline app/views/devise/sessions/new.html.erb]: We don't need email, password etc. We only need authentication via FB.

File: app/views/devise/shared/_links.html.erb [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+
+<%- if devise_mapping.confirmable? && controller_name != 'confirmations' %>
+  <%= link_to "Didn't receive confirmation instructions?", new_confirmation_path(resource_name) %><br />
+<% end -%>
+
+<%- if devise_mapping.lockable? && resource_class.unlock_strategy_enabled?(:email) && controller_name != 'unlocks' %>
+  <%= link_to "Didn't receive unlock instructions?", new_unlock_path(resource_name) %><br />
+<% end -%>
+
+<%- if devise_mapping.omniauthable? %>
```
Comment (vovka) [inline app/views/devise/shared/_links.html.erb]: We need only oauth link. 
Where this partial is rendered? It should be only on home page.

File: app/views/devise/unlocks/new.html.erb [hunk 25]
```diff
@@ -0,0 +1,16 @@
+<h2>Resend unlock instructions</h2>
```
Comment (vovka) [inline app/views/devise/unlocks/new.html.erb]: We don't need unlocking

File: app/views/home/index.html.erb [hunk 26]
```diff
@@ -0,0 +1,10 @@
+<% if user_signed_in? %>
+	<%= current_user.name%><br>
```
Comment (vovka) [inline app/views/home/index.html.erb]: Wrong indentations. Please, use 2 spaces.

File: app/views/home/index.html.erb [hunk 27]
```diff
@@ -0,0 +1,10 @@
...
+	<img src="<%= current_user.image%>">
+	<br>
+	<%= link_to "Log out", destroy_user_session_path, method: :delete %>
+	<% else%>
+	<%= link_to "Sign in with Facebook", user_omniauth_authorize_path(:facebook) %>
+<%end%>
+
+
```
Comment (vovka) [inline app/views/home/index.html.erb]: Redundant line.

File: app/views/home/index.html.erb (hunk 26 above)
Comment (vovka) [inline app/views/home/index.html.erb]: Please, wrap ERB code with spaces, like this:
`<%= current_user.name%>` - wrong
`<%= current_user.name %>` - correct (a space between `name` and `%>`)

File: app/views/layouts/application.html.erb [hunk 28]
```diff
@@ -1,13 +1,14 @@
 <!DOCTYPE html>
```
Comment (vovka) [inline app/views/layouts/application.html.erb]: Why did you change this file? It wasn't necessary.

File: app/views/users/confirmations/new.html.erb [hunk 29]
```diff
@@ -0,0 +1,16 @@
+<h2>Resend confirmation instructions</h2>
```
Comment (vovka) [inline app/views/users/confirmations/new.html.erb]: We don't need confirmation

File: app/views/users/mailer/confirmation_instructions.html.erb [hunk 30]
```diff
@@ -0,0 +1,5 @@
+<p>Welcome <%= @email %>!</p>
```
Comment (vovka) [inline app/views/users/mailer/confirmation_instructions.html.erb]: We don't need confirmation

File: app/views/users/mailer/password_change.html.erb [hunk 31]
```diff
@@ -0,0 +1,3 @@
+<p>Hello <%= @resource.email %>!</p>
```
Comment (vovka) [inline app/views/users/mailer/password_change.html.erb]: we don't need passwords, hence no need to change them.

File: app/views/users/mailer/reset_password_instructions.html.erb [hunk 32]
```diff
@@ -0,0 +1,8 @@
+<p>Hello <%= @resource.email %>!</p>
```
Comment (vovka) [inline app/views/users/mailer/reset_password_instructions.html.erb]: We don't need password reset

File: app/views/users/mailer/unlock_instructions.html.erb [hunk 33]
```diff
@@ -0,0 +1,7 @@
+<p>Hello <%= @resource.email %>!</p>
```
Comment (vovka) [inline app/views/users/mailer/unlock_instructions.html.erb]: we don;t need locking

File: app/views/users/passwords/edit.html.erb [hunk 34]
```diff
@@ -0,0 +1,25 @@
+<h2>Change your password</h2>
```
Comment (vovka) [inline app/views/users/passwords/edit.html.erb]: 1. We don' need passwords.
1. Doesn't it duplicate app/views/devise/passwords/edit.html.erb?

File: app/views/users/passwords/new.html.erb [hunk 35]
```diff
@@ -0,0 +1,16 @@
+<h2>Forgot your password?</h2>
```
Comment (vovka) [inline app/views/users/passwords/new.html.erb]: We don't need passwords

File: app/views/users/registrations/edit.html.erb [hunk 36]
```diff
@@ -0,0 +1,39 @@
+<h2>Edit <%= resource_name.to_s.humanize %></h2>
```
Comment (vovka) [inline app/views/users/registrations/edit.html.erb]: We don't need emails, passwords etc. just FB auth

File: app/views/users/registrations/new.html.erb [hunk 37]
```diff
@@ -0,0 +1,29 @@
+<h2>Sign up</h2>
```
Comment (vovka) [inline app/views/users/registrations/new.html.erb]: We don't need passwords

File: app/views/users/sessions/new.html.erb [hunk 38]
```diff
@@ -0,0 +1,26 @@
+<h2>Log in</h2>
```
Comment (vovka) [inline app/views/users/sessions/new.html.erb]: We don't need emails, passwords etc. just FB auth

File: app/views/users/shared/_links.html.erb [hunk 39]
```diff
@@ -0,0 +1,25 @@
+<%- if controller_name != 'sessions' %>
```
Comment (vovka) [inline app/views/users/shared/_links.html.erb]: 1. We don' need passwords.
2. Doesn't it duplicate app/views/devise/...?

File: app/views/users/shared/_links.html.erb [hunk 40]
```diff
@@ -0,0 +1,25 @@
...
+
+<%- if devise_mapping.confirmable? && controller_name != 'confirmations' %>
+  <%= link_to "Didn't receive confirmation instructions?", new_confirmation_path(resource_name) %><br />
+<% end -%>
+
+<%- if devise_mapping.lockable? && resource_class.unlock_strategy_enabled?(:email) && controller_name != 'unlocks' %>
+  <%= link_to "Didn't receive unlock instructions?", new_unlock_path(resource_name) %><br />
+<% end -%>
+
+<%- if devise_mapping.omniauthable? %>
```
Comment (vovka) [inline app/views/users/shared/_links.html.erb]: See question for app/views/devise/shared/_links.html.erb

File: app/views/users/unlocks/new.html.erb [hunk 41]
```diff
@@ -0,0 +1,16 @@
+<h2>Resend unlock instructions</h2>
```
Comment (vovka) [inline app/views/users/unlocks/new.html.erb]: We don't need locking

**Comment by vovka** [comment]
There are too many unnecessary files. Please, check requirements and clean up.

File: app/controllers/home_controller.rb (hunk 6 above)
Comment (sergiy17) [inline app/controllers/home_controller.rb]: Yes we need this controller because this is the home page for our aplication

File: app/controllers/application_controller.rb [hunk 42]
```diff
@@ -1,8 +1,4 @@
 class ApplicationController < ActionController::Base
-  # Prevent CSRF attacks by raising an exception.
-  # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
-  def hello
-  	render text: "hi from Mars!"
-  end
+  protected
```
Comment (vovka) [inline app/controllers/application_controller.rb]: Whjat this `protected` is?

File: app/controllers/home_controller.rb (hunk 6 above)
Comment (vovka) [inline app/controllers/home_controller.rb]: ok

File: app/models/user.rb [hunk 43]
```diff
@@ -0,0 +1,13 @@
+class User < ActiveRecord::Base
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :recoverable, :rememberable, :trackable, :validatable, :omniauthable, :omniauth_providers => [:facebook]
```
Comment (vovka) [inline app/models/user.rb]: Please, double check if we really need each of these strategies.

File: app/views/home/index.html.erb [hunk 44]
```diff
@@ -0,0 +1,8 @@
+<% if user_signed_in? %>
+  <%= current_user.name %><br>
+  <img src="<%= current_user.image %>">
+  <br>
+  <%= link_to "Log out", destroy_user_session_path, method: :delete %>
+  <% else %>
```
Comment (vovka) [inline app/views/home/index.html.erb]: Please, align this `else` as `if` is.

File: config/initializers/devise.rb [hunk 45]
```diff
@@ -0,0 +1,14 @@
+Devise.setup do |config|
+  config.omniauth :facebook, "1706833466223396", "fe92b6e4f3dd0ef2fefe15d63f92d395",
```
Comment (vovka) [inline config/initializers/devise.rb]: We must not commit secured information. Please, find out how to manage it, key words are `rails secrets`. Respect heroku and codeship environments.

File: config/locales/devise.en.yml [hunk 46]
```diff
@@ -0,0 +1,62 @@
+# Additional translations at https://github.com/plataformatec/devise/wiki/I18n
```
Comment (vovka) [inline config/locales/devise.en.yml]: It is a copy of https://github.com/plataformatec/devise/blob/master/config/locales/en.yml file, isn't it? Can we survive without this duplicate? Just leave it inside the devise gem?

File: app/controllers/users/omniauth_callbacks_controller.rb (hunk 11 above)
Comment (sergiy17) [inline app/controllers/users/omniauth_callbacks_controller.rb]: done

File: config/routes.rb [hunk 47]
```diff
@@ -1,9 +1,12 @@
 Rails.application.routes.draw do
+  get 'home/index'
```
Comment (sergiy17) [inline config/routes.rb]: Prefer single-quoted strings when you don't need string interpolation or special symbols.
  get "home/index"

File: config/environments/development.rb [hunk 48]
```diff
@@ -1,6 +1,6 @@
 Rails.application.configure do
   # Settings specified here will take precedence over those in config/application.rb.
-
+  config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }
```
Comment (sergiy17) [inline config/environments/development.rb]: done

File: db/migrate/20160316195045_add_name_and_image_to_users.rb [hunk 49]
```diff
@@ -0,0 +1,6 @@
+class AddNameAndImageToUsers < ActiveRecord::Migration
+  def change
+  	add_column :users , :name , :string
+  	add_column :users , :image , :string
```
Comment (sergiy17) [inline db/migrate/20160316195045_add_name_and_image_to_users.rb]: done

File: db/migrate/20160316195045_add_name_and_image_to_users.rb [hunk 50]
```diff
@@ -0,0 +1,6 @@
+class AddNameAndImageToUsers < ActiveRecord::Migration
+  def change
+  	add_column :users , :name , :string
```
Comment (sergiy17) [inline db/migrate/20160316195045_add_name_and_image_to_users.rb]: done

File: db/migrate/20160316162024_devise_create_users.rb [hunk 51]
```diff
@@ -0,0 +1,42 @@
...
+      # t.datetime :confirmed_at
+      # t.datetime :confirmation_sent_at
+      # t.string   :unconfirmed_email # Only if using reconfirmable
+
+      ## Lockable
+      # t.integer  :failed_attempts, default: 0, null: false # Only if lock strategy is :failed_attempts
+      # t.string   :unlock_token # Only if unlock strategy is :email or :both
+      # t.datetime :locked_at
+
+
```
Comment (sergiy17) [inline db/migrate/20160316162024_devise_create_users.rb]: done

File: db/migrate/20160316162024_devise_create_users.rb [hunk 52]
```diff
@@ -0,0 +1,42 @@
...
+      t.string   :last_sign_in_ip
+
+      ## Confirmable
+      # t.string   :confirmation_token
+      # t.datetime :confirmed_at
+      # t.datetime :confirmation_sent_at
+      # t.string   :unconfirmed_email # Only if using reconfirmable
+
+      ## Lockable
+      # t.integer  :failed_attempts, default: 0, null: false # Only if lock strategy is :failed_attempts
```
Comment (sergiy17) [inline db/migrate/20160316162024_devise_create_users.rb]: done

File: config/routes.rb [hunk 53]
```diff
@@ -1,9 +1,12 @@
...
+  devise_for :users, :controllers => { :omniauth_callbacks => "users/omniauth_callbacks" }
+
```
Comment (sergiy17) [inline config/routes.rb]: done

File: config/routes.rb [hunk 54]
```diff
@@ -1,9 +1,12 @@
 Rails.application.routes.draw do
+  get 'home/index'
+  devise_for :users, :controllers => { :omniauth_callbacks => "users/omniauth_callbacks" }
```
Comment (sergiy17) [inline config/routes.rb]: done

File: config/initializers/devise.rb [hunk 55]
```diff
@@ -0,0 +1,14 @@
...
+                callback_url: "http://localhost:3000/users/auth/facebook/callback"
+  require 'devise/orm/active_record'
```
Comment (sergiy17) [inline config/initializers/devise.rb]: done

File: app/models/user.rb [hunk 56]
```diff
@@ -0,0 +1,14 @@
+class User < ActiveRecord::Base
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable, :omniauthable, :omniauth_providers => [:facebook]
+  def self.from_omniauth(auth)
+    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
+      user.email = auth.info.email
+      user.password = Devise.friendly_token[0,20]
```
Comment (sergiy17) [inline app/models/user.rb]: done

File: app/models/user.rb [hunk 57]
```diff
@@ -0,0 +1,14 @@
+class User < ActiveRecord::Base
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable, :omniauthable, :omniauth_providers => [:facebook]
+  def self.from_omniauth(auth)
+    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
+      user.email = auth.info.email
```
Comment (sergiy17) [inline app/models/user.rb]: done

File: app/models/user.rb [hunk 58]
```diff
@@ -0,0 +1,14 @@
+class User < ActiveRecord::Base
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable, :omniauthable, :omniauth_providers => [:facebook]
```
Comment (sergiy17) [inline app/models/user.rb]: done

File: app/models/user.rb [hunk 59]
```diff
@@ -0,0 +1,14 @@
...
+  def self.from_omniauth(auth)
+    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
+      user.email = auth.info.email
+      user.password = Devise.friendly_token[0,20]
+      user.name = auth.info.name   # assuming the user model has a name
+      user.image = auth.info.image # assuming the user model has an image
+  end
+end
```
Comment (sergiy17) [inline app/models/user.rb]: done

File: app/models/user.rb [hunk 60]
```diff
@@ -0,0 +1,14 @@
...
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable, :omniauthable, :omniauth_providers => [:facebook]
+  def self.from_omniauth(auth)
+    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
+      user.email = auth.info.email
+      user.password = Devise.friendly_token[0,20]
+      user.name = auth.info.name   # assuming the user model has a name
+      user.image = auth.info.image # assuming the user model has an image
+  end
```
Comment (sergiy17) [inline app/models/user.rb]: done

File: config/routes.rb [hunk 61]
```diff
@@ -1,9 +1,15 @@
 Rails.application.routes.draw do
+  get 'home/index'
+  devise_for :users, :skip => [:password, :sessions], :controllers => { :omniauth_callbacks => "users/omniauth_callbacks" }
```
Comment (sergiy17) [inline config/routes.rb]: done

File: app/models/user.rb [hunk 62]
```diff
@@ -0,0 +1,13 @@
+class User < ActiveRecord::Base
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :recoverable, :rememberable, :trackable, :validatable, :omniauthable, :omniauth_providers => [:facebook]
+  def self.from_omniauth(auth)
+    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
+      user.email = auth.info.email
+      user.password = Devise.friendly_token[0,20]
```
Comment (sergiy17) [inline app/models/user.rb]: done

File: app/models/user.rb [hunk 63]
```diff
@@ -0,0 +1,13 @@
+class User < ActiveRecord::Base
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :recoverable, :rememberable, :trackable, :validatable, :omniauthable, :omniauth_providers => [:facebook]
+  def self.from_omniauth(auth)
+    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
+      user.email = auth.info.email
```
Comment (sergiy17) [inline app/models/user.rb]: done

File: app/models/user.rb (hunk 43 above)
Comment (sergiy17) [inline app/models/user.rb]: i used the new Ruby syntax
was
**my_hash = { :key => 'value' }
**become**
my_hash = { key: 'value' }**

File: app/models/user.rb [hunk 64]
```diff
@@ -0,0 +1,13 @@
...
+      user.password = Devise.friendly_token[0,20]
+      user.name = auth.info.name   # assuming the user model has a name
+      user.image = auth.info.image # assuming the user model has an image
+  end
+end
```
Comment (sergiy17) [inline app/models/user.rb]: done

File: app/views/home/index.html.erb (hunk 44 above)
Comment (sergiy17) [inline app/views/home/index.html.erb]: done

File: config/locales/devise.en.yml (hunk 46 above)
Comment (sergiy17) [inline config/locales/devise.en.yml]: i deleted devise.en.yml

File: app/controllers/application_controller.rb (hunk 42 above)
Comment (sergiy17) [inline app/controllers/application_controller.rb]: This make protected/private methods to be hidden from the outside.
But now you are right, we don't need `protected`

File: db/migrate/20160316162024_devise_create_users.rb [hunk 65]
```diff
@@ -0,0 +1,42 @@
...
+      t.string   :last_sign_in_ip
+
+      ## Confirmable
+      # t.string   :confirmation_token
+      # t.datetime :confirmed_at
+      # t.datetime :confirmation_sent_at
+      # t.string   :unconfirmed_email # Only if using reconfirmable
+
+      ## Lockable
+      # t.integer  :failed_attempts, default: 0, null: false
```
Comment (sergiy17) [inline db/migrate/20160316162024_devise_create_users.rb]: done

File: config/routes.rb [hunk 66]
```diff
@@ -1,9 +1,17 @@
 Rails.application.routes.draw do
-  # The priority is based upon order of creation: first created -> highest priority.
+  get "home/index"
+  devise_for :users, skip: [:password, :sessions],
+    controllers: { omniauth_callbacks: 'users/omniauth_callbacks' }
+
+  devise_scope :user do
+    delete 'sign_out', to: 'devise/sessions#destroy', as: :destroy_user_session
```
Comment (sergiy17) [inline config/routes.rb]: done

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 67]
```diff
@@ -0,0 +1,18 @@
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def facebook
+    # You need to implement the method below in your model (e.g. app/models/user.rb)
+    @user = User.from_omniauth(request.env["omniauth.auth"])
+
+    if @user.persisted?
+      sign_in_and_redirect @user, :event => :authentication #this will throw if @user is not activated
+      set_flash_message(:notice, :success, :kind => "Facebook") if is_navigational_format?
```
Comment (sergiy17) [inline app/controllers/users/omniauth_callbacks_controller.rb]: done

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 68]
```diff
@@ -0,0 +1,18 @@
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def facebook
+    # You need to implement the method below in your model (e.g. app/models/user.rb)
+    @user = User.from_omniauth(request.env["omniauth.auth"])
+
+    if @user.persisted?
+      sign_in_and_redirect @user, :event => :authentication #this will throw if @user is not activated
```
Comment (sergiy17) [inline app/controllers/users/omniauth_callbacks_controller.rb]: done

File: config/routes.rb [hunk 69]
```diff
@@ -1,9 +1,17 @@
 Rails.application.routes.draw do
-  # The priority is based upon order of creation: first created -> highest priority.
+  get "home/index"
+  devise_for :users, skip: [:password, :sessions],
+    controllers: { omniauth_callbacks: 'users/omniauth_callbacks' }
```
Comment (sergiy17) [inline config/routes.rb]: done

File: app/models/user.rb [hunk 70]
```diff
@@ -0,0 +1,14 @@
+class User < ActiveRecord::Base
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :recoverable, :rememberable, :trackable,
+  :validatable, :omniauthable, omniauth_providers: [:facebook]
```
Comment (sergiy17) [inline app/models/user.rb]: done

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 71]
```diff
@@ -0,0 +1,18 @@
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def facebook
+    # You need to implement the method below in your model (e.g. app/models/user.rb)
```
Comment (sergiy17) [inline app/controllers/users/omniauth_callbacks_controller.rb]: done

File: app/views/layouts/application.html.erb (hunk 28 above)
Comment (vovka) [inline app/views/layouts/application.html.erb]: Why did you change this file? It didn't require any change.

Comment (anatoliiibrahimov) [inline app/views/layouts/application.html.erb]: done

**Comment by vovka** [comment]
:+1:

File: config/initializers/devise.rb [hunk 72]
```diff
@@ -0,0 +1,16 @@
+Devise.setup do |config|
+  config.omniauth :facebook, Rails.application.secrets.facebook_app, Rails.application.secrets.facebook_app_key,
+  callback_url: 'http://localhost:3000/users/auth/facebook/callback'
```
Comment (vovka) [inline config/initializers/devise.rb]: Odd code alignment

**Comment by vovka** [comment]
Address hound and my comments

**Comment by vovka** [comment]
Hey guys, remove my commit from this PR
